libraries/AC_PID/AC_PID_2D.h

```diff
@@ -7,87 +7,95 @@
+    //  the derivative is then calculated and filtered
+    //  the integral is then updated based on the setting of the limit flag
+    Vector2f update_all(float target_x, float target_y, Vector2f measurement, bool limit = false);
+    Vector2f update_all(float target_x, float target_y, Vector3f measurement, bool limit = false);
+
+    //  update_i - update the integral
+    //  if the limit flag is set the integral is only allowed to shrink
+    void update_i(bool limit);
 
     // get_pi - get results from pid controller
-    Vector2f    get_pid();
-    Vector2f    get_p() const;
-    Vector2f    get_i();
-    Vector2f    get_i_shrink();   // get_i but do not allow integrator to grow (it may shrink)
-    Vector2f    get_d();
+    Vector2f get_p() const;
+    Vector2f get_i() const;
+    Vector2f get_d() const;
+    Vector2f get_ff();
+    Vector2f get_error() const { return _error; }
 
     // reset_I - reset the integrator
     void        reset_I();
 
     // reset_filter - input and D term filter will be reset to the next value provided to set_input()
-    void        reset_filter();
+    void        reset_filter() { _flags._reset_filter = true; }
 
     // load gain from eeprom
     void        load_gains();
 
     // save gain to eeprom
     void        save_gains();
 
+    /// operator function call for easy initialisation
+    void operator()(float p_val, float i_val, float d_val, float ff_val, float imax_val, float input_filt_E_hz, float input_filt_D_hz, float dt);
+
     // get accessors
-    AP_Float   &kP() { return _kp; }
-    AP_Float   &kI() { return _ki; }
-    float       imax() const { return _imax.get(); }
-    float       filt_hz() const { return _filt_hz.get(); }
-    float       get_filt_alpha() const { return _filt_alpha; }
-    float       filt_d_hz() const { return _filt_d_hz.get(); }
-    float       get_filt_alpha_D() const { return _filt_alpha_d; }
+    AP_Float &kP() { return _kp; }
+    AP_Float &kI() { return _ki; }
+    AP_Float &kD() { return _kd; }
+    AP_Float &ff() { return _kff;}
+    AP_Float &filt_E_hz() { return _filt_E_hz; }
+    AP_Float &filt_D_hz() { return _filt_D_hz; }
+    float imax() const { return _kimax.get(); }
+    float get_filt_alpha(float filt_hz) const;
+    float get_filt_E_alpha() const;
+    float get_filt_D_alpha() const;
 
     // set accessors
-    void        kP(const float v) { _kp.set(v); }
-    void        kI(const float v) { _ki.set(v); }
-    void        kD(const float v) { _kd.set(v); }
-    void        imax(const float v) { _imax.set(fabsf(v)); }
-    void        filt_hz(const float v);
-    void        filt_d_hz(const float v);
-
-    Vector2f    get_integrator() const { return _integrator; }
-    void        set_integrator(const Vector2f &i) { _integrator = i; }
-    void        set_integrator(const Vector3f &i) { _integrator.x = i.x; _integrator.y = i.y; }
+    void kP(const float v) { _kp.set(v); }
+    void kI(const float v) { _ki.set(v); }
```

Comment:
```suggestion
    void set_kP(const float v) { _kp.set(v); }
    void set_kI(const float v) { _ki.set(v); }
```

etc

Reply:
I think I'll leave these names as-is so they are consistent with existing classes... just too much renaming going on already.